File: python/0031.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        def reverse(a,i,j):
            while i<j:
                a[i],a[j]=a[j],a[i]
                i+=1
                j-=1
        i=len(nums)-1
        while i>0 and nums[i]<=nums[i-1]:
            i-=1
        if i==0:
            reverse(nums,0,len(nums)-1)
            return
        reverse(nums,i,len(nums)-1)
        j=len(nums)-1
        while nums[j]>nums[i-1]:
            j-=1
        nums[i-1],nums[j+1]=nums[j+1],nums[i-1]

File: rust/src/bin/0031.rs
struct Solution;

use std::mem::swap;

impl Solution {
    pub fn next_permutation(nums: &mut Vec<i32>) {
        if nums.len() == 0 {
            return;
        }
        let mut i = nums.len() - 1;
        while i > 0 && nums[i] <= nums[i - 1] {
            i -= 1;
        }
        if i == 0 {
            nums.reverse();
            return;
        }
        let target = nums[i - 1];
        let mut l = nums.len() - 1;
        while l > i && nums[l] <= target {
            l -= 1;
        }
        nums.swap(l, i - 1);
        let mut j = nums.len() - 1;
        while j > i {
            nums.swap(i, j);
            i += 1;
            j -= 1;
        }
    }
}


fn main() {
    let mut a = vec!(1, 1, 5);
    for i in 0..7 {
        println!("{:?}", a);
        Solution::next_permutation(&mut a);
    }
    println!("{:?}", a);
}